File: UVa/Contents/d442/d442.c
#include<stdio.h>

int next(int);
int isHappy(int);

int main(void)
{
	int n, temp;
	
	while(scanf("%d", &n) != EOF)
	{
		for(int i = 0; i < n; i++)
		{
			scanf("%d", &temp);
			
			printf("Case #%d: %d is ", i + 1, temp);
			if(isHappy(temp))
				printf("a Happy number.\n");
			else
				printf("an Unhappy number.\n");
		}
	}
	
	return 0;
		
}

int next(int num)
{
    int result = 0;
    int rem;
    
    while(num != 0)
    {
        rem = num % 10;
        num /= 10;
        result += rem * rem;
    }
    
    return result;
}

int isHappy(int n)
{
    int num1 = n;
    int num2 = n;
    
    do
    {
        num1 = next(num1);
        num2 = next(next(num2));
    }while(num2 != num1);
    
    return (num1 == 1) ? 1 : 0;
}

File: UVa/Contents/d442/d442.py
import sys

def nextNum(num):
    result = 0

    while(num != 0):
        rem = num % 10
        num //= 10
        result += rem * rem

    return result

def isHappy(n):    
    num1 = n
    num2 = n
    
    while True:
        num1 = nextNum(num1)
        num2 = nextNum(nextNum(num2))
        
        if num1 == num2:
            break
            
    return num1 == 1

for n in sys.stdin:
    n = int(n)

    for i in range(n):
        num = int(input())
        if isHappy(num):
            print("Case #%d: %d is a Happy number." % (i + 1, num))
        else:
            print("Case #%d: %d is an Unhappy number." % (i + 1, num))
